File: eclipse/plugin/src/test/java/com/adityakumar/engine/DisplayRowModelTest.java
package com.adityakumar.engine;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * Unit tests for {@link DisplayRowModel} — the fold/view-mode display model.
 *
 * <p>Mirrors the canonical test vectors from the VS Code
 * {@code extension/test/rowModel.test.ts} file.
 *
 * <p>Status encoding: u=0 (unchanged), a=1 (added), r=2 (removed), c=3 (changed).
 */
class DisplayRowModelTest {

    // ------------------------------------------------------------------ //
    //  Helpers                                                             //
    // ------------------------------------------------------------------ //

    /** Parse a compact spec like "uuacuuruu" into a byte[] status column. */
    private static byte[] statuses(String spec) {
        byte[] result = new byte[spec.length()];
        for (int i = 0; i < spec.length(); i++) {
            result[i] = switch (spec.charAt(i)) {
                case 'u' -> 0;
                case 'a' -> 1;
                case 'r' -> 2;
                case 'c' -> 3;
                default  -> throw new IllegalArgumentException("bad char: " + spec.charAt(i));
            };
        }
        return result;
    }

    /** Build a model in ALL mode with an empty expanded set. */
    private static DisplayRowModel.Model all(String spec) {
        return DisplayRowModel.build(statuses(spec), DisplayRowModel.ViewMode.ALL, 3, Set.of());
    }

    /** Build a model in CHANGES mode. */
    private static DisplayRowModel.Model changes(String spec) {
        return DisplayRowModel.build(statuses(spec), DisplayRowModel.ViewMode.CHANGES, 3, Set.of());
    }

    /** Build a model in COLLAPSED mode with context=3 and no expanded runs. */
    private static DisplayRowModel.Model collapsed(String spec) {
        return DisplayRowModel.build(statuses(spec), DisplayRowModel.ViewMode.COLLAPSED, 3, Set.of());
    }

    /** Build a model in COLLAPSED mode with context=3 and some expanded run-starts. */
    private static DisplayRowModel.Model collapsed(String spec, Integer... expandedRunStarts) {
        return DisplayRowModel.build(statuses(spec), DisplayRowModel.ViewMode.COLLAPSED, 3,
                new HashSet<>(Arrays.asList(expandedRunStarts)));
    }

    // ------------------------------------------------------------------ //
    //  ALL mode                                                            //
    // ------------------------------------------------------------------ //

    @Test
    void allModeIsIdentityNoAllocation() {
        var m = all("uacru");
        assertEquals(5, m.count);
        assertNull(m.map,            "ALL mode: map must be null");
        assertNull(m.absToDisplay,   "ALL mode: absToDisplay must be null");
        assertNull(m.displayStatuses,"ALL mode: displayStatuses must be null");
        assertEquals(0, m.folds.length);
    }

    @Test
    void allModeDisplayOfIsIdentity() {
        var m = all("uacru");
        for (int i = 0; i < 5; i++) {
            assertEquals(i, m.displayOf(i));
        }
    }

    @Test
    void emptyInputIsIdentityRegardlessOfMode() {
        var m = DisplayRowModel.build(new byte[0], DisplayRowModel.ViewMode.CHANGES, 3, Set.of());
        assertEquals(0, m.count);
        assertNull(m.map);
    }

    // ------------------------------------------------------------------ //
    //  CHANGES mode                                                        //
    // ------------------------------------------------------------------ //

    @Test
    void changesKeepsOnlyDiffRows() {
        // "uuacuuruu" — changes at abs 2(a), 3(c), 6(r)
        var m = changes("uuacuuruu");
        assertEquals(3, m.count);
        assertNotNull(m.map);
        assertArrayEquals(new int[]{2, 3, 6}, m.map);
    }

    @Test
    void changesDisplayStatuses() {
        var m = changes("uuacuuruu");
        assertNotNull(m.displayStatuses);
        assertArrayEquals(new byte[]{1, 3, 2}, m.displayStatuses);
    }

    @Test
    void changesAbsToDisplayHiddenRows() {
        // "uuacuuruu": changes at 2(d=0), 3(d=1), 6(d=2)
        // abs 0,1 (unchanged before first change) -> d=0 (next visible)
        // abs 2   (a) -> d=0
        // abs 3   (c) -> d=1
        // abs 4,5 (unchanged) -> d=2 (next visible is abs 6 at d=2)
        // abs 6   (r) -> d=2
        // abs 7,8 (trailing unchanged) -> d=2 (clamped to count-1=2)
        var m = changes("uuacuuruu");
        assertArrayEquals(new int[]{0, 0, 0, 1, 2, 2, 2, 2, 2}, m.absToDisplay);
    }

    @Test
    void changesAllUnchangedIsEmptyList() {
        var m = changes("uuuu");
        assertEquals(0, m.count);
    }

    @Test
    void changesAbsToDisplayForAllUnchanged() {
        // count=0, so all hidden rows clamp to max(0, -1) = 0
        var m = changes("uuuu");
        assertNotNull(m.absToDisplay);
        assertArrayEquals(new int[]{0, 0, 0, 0}, m.absToDisplay);
    }

    // ------------------------------------------------------------------ //
    //  COLLAPSED mode                                                      //
    // ------------------------------------------------------------------ //

    @Test
    void collapsedFoldsLongUnchangedRun() {
        // "uuuuuuuuuuc": 10 unchanged then 1 changed.
        // context=3: show rows 0,1,2; fold [3,7); show rows 7,8,9; show row 10.
        var m = collapsed("uuuuuuuuuuc");
        assertEquals(8, m.count);  // 3 + 1 fold + 3 + 1 changed
        assertEquals(1, m.folds.length);

        var fold = m.folds[0];
        assertEquals(0, fold.runStart);
        assertEquals(3, fold.start);
        assertEquals(7, fold.end);
        assertEquals(4, fold.count);

        // map: [0,1,2, -1 (fold_id=0), 7,8,9, 10]
        assertNotNull(m.map);
        assertArrayEquals(new int[]{0, 1, 2, -1, 7, 8, 9, 10}, m.map);
    }

    @Test
    void collapsedAbsToDisplayForFolded() {
        // "uuuuuuuuuuc": fold slot at display index 3.
        // Rows 3,4,5,6 are hidden -> all map to d=3.
        var m = collapsed("uuuuuuuuuuc");
        assertNotNull(m.absToDisplay);
        // Visible rows 0,1,2 -> d=0,1,2
        assertEquals(0, m.absToDisplay[0]);
        assertEquals(1, m.absToDisplay[1]);
        assertEquals(2, m.absToDisplay[2]);
        // Hidden rows 3..6 -> d=3 (the fold slot)
        assertEquals(3, m.absToDisplay[3]);
        assertEquals(3, m.absToDisplay[4]);
        assertEquals(3, m.absToDisplay[5]);
        assertEquals(3, m.absToDisplay[6]);
        // Visible rows 7,8,9 -> d=4,5,6
        assertEquals(4, m.absToDisplay[7]);
        assertEquals(5, m.absToDisplay[8]);
        assertEquals(6, m.absToDisplay[9]);
        // Changed row 10 -> d=7
        assertEquals(7, m.absToDisplay[10]);
    }

    @Test
    void collapsedDoesNotFoldRunOfExactly2xContext() {
        // "uuuuuuc": 6 unchanged = 2*3, then change. Must NOT fold.
        var m = collapsed("uuuuuuc");
        assertEquals(0, m.folds.length);
        assertEquals(7, m.count);
        // All 7 display slots are identity-like
        assertNotNull(m.map);
        assertArrayEquals(new int[]{0, 1, 2, 3, 4, 5, 6}, m.map);
    }

    @Test
    void collapsedFoldsRunOf2xContextPlusOne() {
        // "uuuuuuuc": 7 unchanged = 2*3+1, then change. MUST fold, hiding exactly 1 row.
        var m = collapsed("uuuuuuuc");
        assertEquals(1, m.folds.length);
        var fold = m.folds[0];
        assertEquals(3, fold.start);
        assertEquals(4, fold.end);   // runEnd=7, context=3 -> end=7-3=4
        assertEquals(1, fold.count);
    }

    @Test
    void collapsedExpandedRunShowsAllRows() {
        // "uuuuuuuuuuc" with runStart=0 expanded.
        var m = collapsed("uuuuuuuuuuc", 0);
        assertEquals(0, m.folds.length);
        assertEquals(11, m.count);  // all 10 unchanged + 1 changed
    }

    @Test
    void collapsedTwoIndependentFolds() {
        // "uuuuuuuucuuuuuuuuc": run A [0,8), changed @8, run B [9,17), changed @17
        // context=3: show 0,1,2; fold[3,5); show 5,6,7; show 8;
        //            show 9,10,11; fold[12,14); show 14,15,16; show 17
        var m = collapsed("uuuuuuuucuuuuuuuuc");
        assertEquals(2, m.folds.length);

        var foldA = m.folds[0];
        assertEquals(0, foldA.runStart);
        assertEquals(3, foldA.start);
        assertEquals(5, foldA.end);
        assertEquals(2, foldA.count);

        var foldB = m.folds[1];
        assertEquals(9, foldB.runStart);
        assertEquals(12, foldB.start);
        assertEquals(14, foldB.end);
        assertEquals(2, foldB.count);
    }

    @Test
    void collapsedFoldMarkerHasStatus0InDisplayStatuses() {
        var m = collapsed("uuuuuuuuuuc");
        assertNotNull(m.displayStatuses);
        // Slot 3 is the fold marker -> status 0
        assertEquals(0, m.displayStatuses[3]);
        // Slot 7 is the changed row -> status 3
        assertEquals(3, m.displayStatuses[7]);
    }

    @Test
    void collapsedDisplayOfDecoding() {
        var m = collapsed("uuuuuuuuuuc");
        // displayOf wraps absToDisplay
        assertEquals(3, m.displayOf(4));   // row 4 is in the fold at d=3
        assertEquals(7, m.displayOf(10));  // changed row 10 at d=7
    }

    @Test
    void collapsedAbsoluteOfDecoding() {
        var m = collapsed("uuuuuuuuuuc");
        // Display slot 0 -> abs row 0
        assertEquals(0, m.absoluteOf(0));
        // Display slot 3 -> fold (returns -1)
        assertEquals(-1, m.absoluteOf(3));
        // Display slot 7 -> abs row 10
        assertEquals(10, m.absoluteOf(7));
    }

    @Test
    void collapsedFoldAtDecoding() {
        var m = collapsed("uuuuuuuuuuc");
        // Slot 3 is a fold
        var fold = m.foldAt(3);
        assertNotNull(fold);
        assertEquals(3, fold.start);
        assertEquals(7, fold.end);
        // Slot 0 is a real row
        assertNull(m.foldAt(0));
    }

    @Test
    void allModeAbsoluteOfAndFoldAt() {
        var m = all("uuuc");
        // identity: absoluteOf(i) == i, foldAt always null
        assertEquals(2, m.absoluteOf(2));
        assertNull(m.foldAt(0));
        assertNull(m.foldAt(3));
    }
}

File: jetbrains/src/test/kotlin/engine/DisplayRowModelTest.kt
package com.adityakumar.engine

import com.adityakumar.engine.DisplayRowModel.ViewMode
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * Unit tests for DisplayRowModel.
 * Mirrors the semantics of extension/test/rowModel.test.ts exactly.
 *
 * Status encoding helper: u=0, a=1, r=2, c=3
 */
class DisplayRowModelTest {

    /** Build a ByteArray status column from a compact string. */
    private fun statuses(spec: String): ByteArray {
        return ByteArray(spec.length) { i ->
            when (spec[i]) {
                'u' -> 0
                'a' -> 1
                'r' -> 2
                'c' -> 3
                else -> throw IllegalArgumentException("bad status char: ${spec[i]}")
            }
        }
    }

    /**
     * Expand the model into the display sequence: absolute row indices and
     * fold descriptors, matching the display() helper in rowModel.test.ts.
     */
    private sealed class DisplayEntry {
        data class Row(val abs: Int) : DisplayEntry()
        data class Fold(val start: Int, val end: Int) : DisplayEntry()
    }

    private fun display(model: DisplayRowModel.Model): List<DisplayEntry> {
        val out = mutableListOf<DisplayEntry>()
        for (d in 0 until model.count) {
            val v = model.map?.get(d) ?: d
            if (model.map != null && v < 0) {
                val fold = model.folds[-1 - v]
                out.add(DisplayEntry.Fold(fold.start, fold.end))
            } else {
                out.add(DisplayEntry.Row(v))
            }
        }
        return out
    }

    private fun row(abs: Int) = DisplayEntry.Row(abs)
    private fun fold(start: Int, end: Int) = DisplayEntry.Fold(start, end)

    // ── ALL mode ───────────────────────────────────────────────────────────────

    @Test fun `all mode is identity regardless of content`() {
        val s = statuses("uacru")
        val m = DisplayRowModel.build(s, ViewMode.ALL)
        assertEquals(5, m.count)
        assertNull(m.map)
        assertNull(m.absToDisplay)
        assertNull(m.displayStatuses)
        assertEquals(emptyList<DisplayRowModel.Fold>(), m.folds)
    }

    @Test fun `all mode empty input is identity`() {
        val m = DisplayRowModel.build(ByteArray(0), ViewMode.ALL)
        assertEquals(0, m.count)
        assertNull(m.map)
    }

    @Test fun `empty input in changes mode is identity`() {
        val m = DisplayRowModel.build(ByteArray(0), ViewMode.CHANGES)
        assertEquals(0, m.count)
        assertNull(m.map)
    }

    // ── CHANGES mode ───────────────────────────────────────────────────────────

    @Test fun `changes mode keeps only rows that differ in order`() {
        // "uuacuuruu" → changes at abs 2(a), 3(c), 6(r)
        val s = statuses("uuacuuruu")
        val m = DisplayRowModel.build(s, ViewMode.CHANGES)
        assertEquals(listOf(row(2), row(3), row(6)), display(m))
        assertArrayEquals(byteArrayOf(1, 3, 2), m.displayStatuses)
    }

    @Test fun `changes mode maps hidden rows to next visible display row`() {
        val s = statuses("uuacuuruu") // changes at 2,3,6
        val m = DisplayRowModel.build(s, ViewMode.CHANGES)
        val a = m.absToDisplay!!
        // abs 0,1 → first visible = d=0
        assertEquals(0, a[0])
        assertEquals(0, a[1])
        // abs 2 (a) → its own display slot d=0
        assertEquals(0, a[2])
        // abs 3 (c) → d=1
        assertEquals(1, a[3])
        // abs 4,5 (unchanged) → next visible is abs 6 at d=2
        assertEquals(2, a[4])
        assertEquals(2, a[5])
        // abs 6 (r) → d=2
        assertEquals(2, a[6])
        // abs 7,8 (trailing unchanged) → clamped to count-1=2
        assertEquals(2, a[7])
        assertEquals(2, a[8])
    }

    @Test fun `changes mode all-unchanged collapses to empty list`() {
        val s = statuses("uuuu")
        val m = DisplayRowModel.build(s, ViewMode.CHANGES)
        assertEquals(0, m.count)
    }

    // ── COLLAPSED mode ─────────────────────────────────────────────────────────

    @Test fun `collapsed folds a long unchanged run with context on each side`() {
        // 10 u then 1 c; context=3 → show 0,1,2, fold [3,7), show 7,8,9, then 10
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        assertEquals(
            listOf(row(0), row(1), row(2), fold(3, 7), row(7), row(8), row(9), row(10)),
            display(m)
        )
        val f = m.folds[0]
        assertEquals(0, f.runStart)
        assertEquals(3, f.start)
        assertEquals(7, f.end)
        assertEquals(4, f.count)
    }

    @Test fun `collapsed does not fold a run of exactly 2 times context`() {
        // 6 u then c; runLen=6 == 2*3 — threshold is >, so NOT folded
        val s = statuses("uuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        assertEquals(0, m.folds.size)
        assertEquals(
            listOf(row(0), row(1), row(2), row(3), row(4), row(5), row(6)),
            display(m)
        )
    }

    @Test fun `collapsed folds a run of 2 times context plus 1`() {
        // 7 u then c; runLen=7 > 2*3=6 → fold [3,4)
        val s = statuses("uuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        assertEquals(1, m.folds.size)
        val f = m.folds[0]
        assertEquals(3, f.start)
        assertEquals(4, f.end)
        assertEquals(1, f.count)
    }

    @Test fun `collapsed expanding a run reveals every row`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3, expanded = setOf(0))
        assertEquals(0, m.folds.size)
        assertEquals(
            (0..10).map { row(it) },
            display(m)
        )
    }

    @Test fun `collapsed folds each long run independently`() {
        // run A (8 u), change, run B (8 u), change
        val s = statuses("uuuuuuuucuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        assertEquals(2, m.folds.size)
        val expected = listOf(
            row(0), row(1), row(2), fold(3, 5), row(5), row(6), row(7),
            row(8),
            row(9), row(10), row(11), fold(12, 14), row(14), row(15), row(16),
            row(17)
        )
        assertEquals(expected, display(m))
    }

    @Test fun `collapsed hidden rows in a fold map to the fold display index`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        // Fold sits at display index 3 (after rows 0,1,2). Rows 3..6 are hidden.
        val a = m.absToDisplay!!
        assertEquals(3, a[3])
        assertEquals(3, a[4])
        assertEquals(3, a[5])
        assertEquals(3, a[6])
        // Visible rows map to their own display index
        assertEquals(4, a[7])
        assertEquals(7, a[10])
    }

    @Test fun `collapsed visible rows map to their own display index`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        val a = m.absToDisplay!!
        assertEquals(0, a[0])
        assertEquals(1, a[1])
        assertEquals(2, a[2])
        assertEquals(4, a[7])
        assertEquals(5, a[8])
        assertEquals(6, a[9])
        assertEquals(7, a[10])
    }

    @Test fun `collapsed fold map entries are negative and decode correctly`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        // Display slot 3 should be the fold
        val mapVal = m.map!![3]
        assertTrue(mapVal < 0, "fold map entry should be negative, got $mapVal")
        val foldId = -1 - mapVal
        assertEquals(0, foldId)
        assertEquals(m.folds[0], m.folds[foldId])
    }

    @Test fun `collapsed displayStatuses fold slot is 0 (unchanged)`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        // Display slot 3 is the fold; its displayStatus must be 0
        assertEquals(0.toByte(), m.displayStatuses!![3])
        // Display slot 7 is the changed row
        assertEquals(3.toByte(), m.displayStatuses!![7])
    }

    // ── displayOf helper ───────────────────────────────────────────────────────

    @Test fun `displayOf returns abs when map is null (all mode)`() {
        val s = statuses("uacru")
        val m = DisplayRowModel.build(s, ViewMode.ALL)
        with(DisplayRowModel) {
            assertEquals(3, m.displayOf(3))
        }
    }

    @Test fun `displayOf uses absToDisplay when present`() {
        val s = statuses("uuacuuruu")
        val m = DisplayRowModel.build(s, ViewMode.CHANGES)
        with(DisplayRowModel) {
            assertEquals(0, m.displayOf(2))  // abs 2 (a) → display 0
            assertEquals(1, m.displayOf(3))  // abs 3 (c) → display 1
            assertEquals(2, m.displayOf(6))  // abs 6 (r) → display 2
        }
    }

    // ── hiddenInFold helper ────────────────────────────────────────────────────

    @Test fun `hiddenInFold returns null for visible rows`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        with(DisplayRowModel) {
            assertNull(m.hiddenInFold(0))   // visible
            assertNull(m.hiddenInFold(10))  // the changed row
        }
    }

    @Test fun `hiddenInFold returns fold for hidden rows`() {
        val s = statuses("uuuuuuuuuuc")
        val m = DisplayRowModel.build(s, ViewMode.COLLAPSED, context = 3)
        with(DisplayRowModel) {
            val fold = m.hiddenInFold(4)
            assertNotNull(fold)
            assertEquals(3, fold!!.start)
            assertEquals(7, fold.end)
        }
    }

    @Test fun `hiddenInFold returns null in all mode (no absToDisplay)`() {
        val s = statuses("uuuuu")
        val m = DisplayRowModel.build(s, ViewMode.ALL)
        with(DisplayRowModel) {
            assertNull(m.hiddenInFold(2))
        }
    }
}
